### PR TITLE
Progress Bar & Content Icon 

### PR DESCRIPTION
**Pull request by christianmemije**
## Summary

* Added a progress bar global component.
* Modified the content-icon and made it a global component.

## Screenshot
image

### CODE REVIEW

**Comment by jamalex** [comment]
Cool! To avoid confusion with the time sliders in the players, it would perhaps be better to put this in the top-right of the window, just above the content, and have it be small (like 100-200px)?

**Comment by christianmemije** [comment]
Sounds good, brb.

**Comment by christianmemije** [comment]
@jamalex Updated. Looked better left aligned.

Updated: 
image

**Comment by rtibbles** [comment]
I think @ralphiee22 would appreciate it if the progress bar was split out into a separate component registered on the global Vue instance (check back to the previous PR where we did that if you don't remember how, @christianmemije!) so that he could use the 'progress-bar' component in his stuff too!

No reason to hold up on the merge here, but perhaps you two could collaborate on it.

**Comment by rtibbles** [comment]
The only other issue is that it seems to have created a lot of white space between the progress bar and the video - could you look into that before merge?

**Comment by christianmemije** [comment]
Addressed those issues ^, ready to merge.

File: kolibri/core/assets/src/vue/progress-bar/index.vue [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<template>
+
+  <div>
+    <div class="visuallyhidden" id="progress-bar-label">{{ $tr('progressComplete') }}</div>
```
Comment (radinamatic) [inline kolibri/core/assets/src/vue/progress-bar/index.vue]: How about `Your progress in this topic is:` as the text for this label?
_(corrected from "video" to "topic" to include other types of content)_

Comment (radinamatic) [inline kolibri/core/assets/src/vue/progress-bar/index.vue]: Or is it supposed to work on other types of content? I see it on PDF content viewer too...

File: kolibri/core/assets/src/vue/progress-bar/index.vue [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    <div class="progress-bar-wrapper"
+      role="progress-bar"
```
Comment (radinamatic) [inline kolibri/core/assets/src/vue/progress-bar/index.vue]: [Role should be `role="progressbar"`](https://www.w3.org/TR/wai-aria/roles#progressbar).

File: kolibri/core/assets/src/vue/progress-bar/index.vue [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      aria-labelledby="progress-bar-label"
+      :aria-valuenow="progressPercent"
+      aria-valuemin="0"
+      aria-valuemax="100">
+      <div class="progress-bar-complete" v-bind:style="{ width: progressPercent + '%' }"></div>
+    </div>
+    <p class="progress-bar-text">{{ progressPercent + ' % ' + $tr('complete') }}</p>
```
Comment (radinamatic) [inline kolibri/core/assets/src/vue/progress-bar/index.vue]: I'm just pondering if we should include `aria-live="polite"` here, but let's hold on it for now until we can test in action!

File: kolibri/core/assets/src/vue/progress-bar/index.vue (hunk 1 above)
Comment (rtibbles) [reply kolibri/core/assets/src/vue/progress-bar/index.vue]: Yes, this should be across all content types.

**Comment by radinamatic** [comment]
I'm not able to see it in action (yet), but I did notice one thing: progressbar in gets loaded first right above the `Download media` button, and then jumps up.

**Comment by radinamatic** [comment]
Screenshot for the above comment:

anim

File: kolibri/core/assets/src/vue/progress-bar/index.vue (hunk 3 above)
Comment (indirectlylit) [reply kolibri/core/assets/src/vue/progress-bar/index.vue]: shouldn't be a space between the number and the `%` char

**Comment by christianmemije** [comment]
@radinamatic That jumping issue will be addressed that I will open as soon as this one is merged in!

**Comment by indirectlylit** [comment]
My personal reaction: I find this bar distracting in a few ways.

* still somewhat large
* the updates keeps attracting my peripheral attention while watching a video
* the percentages shown are very precise (e.g. 47%), but the updates happen sporadically rather than continuously so the values are inaccurate most of the time.

I still feel like it's more appropriate to use a simple 3-state progress indicator (unstarted/started/completed) because:

* this is consistent with the card grid views
* the percentages we're displaying are inaccurate
* it will update less often and therefore be less distracting

**Comment by radinamatic** [comment]
I don't disagree with @indirectlylit's comments, especially on the "distracting" part, and just 3 states sound like a more simple information relayed to the user. 

Had my doubts about the usefulness a detailed percentages display before, but maybe the feature like "Would you like to resume the viewing of this topic from where you left off the last time?" could be a better choice (it is frequently included in standalone video players)? I'm aware that this may involve juggling users session/cookies and other things, and that it may not be the right moment to implement it...

**Comment by christianmemije** [comment]
@indirectlylit Possible Progress Icons:
progress_icons

**Comment by jamalex** [comment]
> Possible Progress Icons:

Ooh, cool! Just to clarify, is the idea that these could actually be more granular %? The data returned for both topics and content nodes is a floating value between 0 and 1, so perhaps it could use CSS to represent the actual progress fraction in there?

**Comment by jamalex** [comment]
> the percentages shown are very precise

Yeah, I don't think there's much value in actually showing the number. But more granular progress indication on there is nice. We do have an interval timer to ensure it's roughly up to date.

**Comment by jamalex** [comment]
> this is consistent with the card grid views

I think it would be nice showing the more granular display on cards as well.

**Comment by rtibbles** [comment]
I think in this case, we should keep the progress bar component - as it will presumably be used by @ralphiee22 in the content import/export app, but embellish the 'content icon' component that currently does this rendering to use CSS for the incremental progress, rather than three different SVG files (and also to take a float value, rather than the three possible categories).

**Comment by jtamiace** [comment]
How feasible is it to show time left, rather than percent completed? i.e. "2:32 left" or even shorten it to "2 minutes left"? This could accompany the progress icon within the content viewer, and could even appear while exploring content or viewing other content cards. 

i'll post some sketches of what i mean in case this doesn't make sense

**Comment by jtamiace** [comment]
I also think the progress bar sort of repeats what the time trackers for audio and video are already doing

**Comment by rtibbles** [comment]
@jtamiace The progress tracker is tracking your completion towards 'completion' of the content, as such, it is cross-content, and also across session.

The time trackers on the video and audio, by their nature, track your current position in the audio or video stream, not how much of the content you have actually engaged with (i.e. jumping to the end of the video does not count as 'completing' that video).

**Comment by christianmemije** [comment]
@indirectlylit  Content-icon on content-page: 
image

File: kolibri/core/assets/src/vue/content-icon/index.vue [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        cx="16"
+        r="13.5"
+        class="inner-circle">
+      </circle>
+      <svg v-if="thisIs('audio')" src="./content-icons/audio.svg" class="content-type-icon"></svg>
+      <svg v-if="thisIs('document')" src="./content-icons/document.svg" class="content-type-icon"></svg>
+      <svg v-if="thisIs('exercise')" src="./content-icons/exercise.svg" class="content-type-icon"></svg>
+      <svg v-if="thisIs('video')" src="./content-icons/video.svg" class="content-type-icon"></svg>
+    </svg>
+  </div>
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/content-icon/index.vue:26]: is it necessary to wrap the `svg` in a `div`?

File: kolibri/core/assets/src/vue/content-icon/index.vue [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+</template>
+
+
+<script>
+
+  const KINDS = ['audio', 'document', 'video']; // not 'exercise' for now
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/content-icon/index.vue]: Let's add this to the core constants file like:

```
...

const ContentKinds = {
  AUDIO: 'audio',
  DOCUMENT: 'document',
  VIDEO: 'video',
};

module.exports = {
  UserKinds,
  ContentKinds,
};

```

Then they can be re-used in e.g. the content-renderer.

File: kolibri/core/assets/src/vue/content-icon/index.vue [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      },
+      size: {
+        type: Number,
+        default: 30,
+      },
+      progress: {
+        type: Number,
+        default: 0.0,
+        validator(value) {
+          return (value >= 0.0) && (value <= 1.0);
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/content-icon/index.vue:58]: nice!

File: kolibri/core/assets/src/vue/progress-bar/index.vue [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    $trNameSpace: 'progressBar',
+    $trs: {
+      yourProgressIs: 'Your progress is:',
+      complete: 'Complete',
+    },
+    props: {
+      progressPercent: {
+        type: Number,
+        default: 0,
+        required: true,
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/progress-bar/index.vue:30]: If something is required, it shouldn't need a default.

Also, would be great to add a validator to this like you have on the other component. Right now it's not obvious whether it takes a value in the range [0, 100] or [0.0, 1.0]

File: kolibri/core/assets/src/vue/content-icon/index.vue (hunk 4 above)
Comment (christianmemije) [reply kolibri/core/assets/src/vue/content-icon/index.vue:26]: Yes

**Comment by christianmemije** [comment]
Updated

File: kolibri/core/assets/src/vue/content-icon/index.vue [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        type: String,
+        required: true,
+        validator(value) {
+          for (const contentKind in ContentKinds) {
+            if (ContentKinds[contentKind] === value) {
+              return true;
+            }
+          }
+          return false;
+        },
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/content-icon/index.vue:71]: can be written concisely as:

``` javascript
validator(thisValue) {
  return ContentKinds.some(kind => kind === thisValue);
}
```

**Comment by indirectlylit** [comment]
looking good! testing it now.

File: kolibri/core/assets/src/vue/content-icon/index.vue (hunk 8 above)
Comment (indirectlylit) [reply kolibri/core/assets/src/vue/content-icon/index.vue:71]: oh - I forgot `ContentKinds` is an object. Might not be that simple then.

**Comment by indirectlylit** [comment]
edges seem to appear aliased in the grid view:

image

File: kolibri/core/assets/src/vue/content-icon/index.vue (hunk 8 above)
Comment (christianmemije) [reply kolibri/core/assets/src/vue/content-icon/index.vue:71]: Ya, I couldn't find a more concise since we are looking at the values of the object.

**Comment by indirectlylit** [comment]
nice job getting it to render in IE9 - I know this stuff is super difficult to get working cross platform.

I think we'll need to find a way to to address the "jaggies", but it will likely require more sophisticated auto-generation of SVG paths rather than the CSS strategies used here.

I created a new card to start us down that path: https://trello.com/c/STjpobDa/572-investigate-combining-d3-helper-functions-with-vue-based-rendering

**Pull request merged**